include/drivers/rtc/maxim_ds3231.h

```diff
@@ -0,0 +1,602 @@
+ *
+ * If clear the alarm fires only when the RTC minutes matches the
+ * alarm minutes.  The bit for IGNSE must be clear.
+ *
+ * If set the alarm minutes field is ignored and alarms will be
+ * triggered based on IGNSE. The bits for IGNHR and IGNDA must both be
+ * set.
+ *
+ * Bit maps to A1M2 or A2M2 and is used in
+ * maxim_ds3231_alarm_configuration::alarm_flags.
+ */
+#define MAXIM_DS3231_ALARM_FLAGS_IGNMN BIT(1)
+
+/** @brief Control alarm behavior on match in hours field.
+ *
+ * If clear the alarm fires only when the RTC hours matches the
+ * alarm hours.  The bits for IGNMN and IGNSE must be clear.
+ *
+ * If set the alarm hours field is ignored and alarms will be
+ * triggered based on IGNMN and IGNSE.  The bit for IGNDA must be set.
+ *
+ * Bit maps to A1M3 or A2M3 and is used in
+ * maxim_ds3231_alarm_configuration::alarm_flags.
+ */
+#define MAXIM_DS3231_ALARM_FLAGS_IGNHR BIT(2)
+
+/** @brief Control alarm behavior on match in day/date field.
+ *
+ * If clear the alarm fires only when the RTC day/date matches the
+ * alarm day/date, mediated by MAXIM_DS3231_ALARM_FLAGS_DAY.  The bits
+ * for IGNHR, IGNMN, and IGNSE must be clear
+ *
+ * If set the alarm day/date field is ignored and an alarm will be
+ * triggered based on IGNHR, IGNMN, and IGNSE.
+ *
+ * Bit maps to A1M4 or A2M4 and is used in
+ * maxim_ds3231_alarm_configuration::alarm_flags.
+ */
+#define MAXIM_DS3231_ALARM_FLAGS_IGNDA BIT(3)
+
+/** @brief Control match on day of week versus day of month
+ *
+ * Set the flag to match on day of week; clear it to match on day of
+ * month.
+ *
+ * Bit maps to DY/DTn in corresponding
+ * maxim_ds3231_alarm_configuration::alarm_flags.
+ */
+#define MAXIM_DS3231_ALARM_FLAGS_DOW BIT(4)
+
+/** @brief Indicates that the alarm should be disabled once it fires.
+ *
+ * Set the flag in the maxim_ds3231_alarm_configuration::alarm_flags
+ * field to cause the alarm to be disabled when the interrupt fires,
+ * prior to invoking the corresponding handler.
+ *
+ * Leave false to allow the alarm to remain enabled so it will fire
+ * again on the next match.
+ */
+#define MAXIM_DS3231_ALARM_FLAGS_AUTODISABLE BIT(7)
+
+/**
+ * @brief RTC DS3231 Driver-Specific API
+ * @defgroup rtc_interface Real Time Clock interfaces
+ * @ingroup io_interfaces
+ * @{
+ */
+
+/** @brief Signature for DS3231 alarm callbacks.
+ *
+ * The alarm callback is invoked from the system work queue thread.
+ * At the point the callback is invoked the corresponding alarm flags
+ * will have been cleared from the device status register.  The
+ * callback is permitted to invoke operations on the device.
+ *
+ * @param dev the device from which the callback originated
+ *
+ * @param syncclock the value from maxim_ds3231_read_syncclock() at the
+ * time the alarm interrupt was processed.
+ *
+ * @param user_data the corresponding parameter from
+ * maxim_ds3231_alarm::user_data.
+ */
+typedef void (*maxim_ds3231_alarm_callback_handler_t)(struct device *dev,
```

Comment:
`maxim_ds3231_alarm_callback_handler_t` is to long, please use shorter names, here `ds3231_alarm_cb_handler` is fine.

Overall just ds3231/DS3231 instead of maxim_ds3231/MAXIM_DS3231 would be sufficient.

Why not generally abbreviate `callback` with `cb` if the prefix give the only feasible interpretation like `ds3231_notify_cb`?

Reply:
While I started with `ds3231_` as the prefix, and that has prior art in tree for (e.g.) `ccs811_`, Zephyr has not carefully managed namespaces and I don't want to assume that there will never be conflicts between device model numbers across vendors.  Standardizing on the compatible name makes sense.  The negative impact is small, only affecting those who are using the custom API.

This proposed policy is something we should discuss more widely.  So: #24978

`callback` is unambiguous while `cb` is only clear in context with interpretation.  IMO the biggest problem with long-but-clear names is a consequence of 8-space indentation with a 78 character line limit.  I have no reason to believe the shorter version ever directly improves comprehension, and some reason to believe it makes it worse (increased cognitive load from requiring an interpretation at all).

I'm seeing 44 `foo_callback_t` typedefs, and 96 `foo_cb_t` typedefs, in Zephyr today.

Not to say I won't change it if people insist, but that's why I picked the names that I did.